runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/DataflowLocationIT.java

```diff
@@ -0,0 +1,120 @@
+
+/**
+ * Tests that the Dataflow runner accepts various --XLocation options with and without trailing '/'.
+ */
+@RunWith(JUnit4.class)
+public class DataflowLocationIT {
+
+  /**
+   * Builds and runs the test using the given {@link Pipeline} We have to invoke the code in this
+   * awkward way so that the calling test can set options after the {@link TestDataflowRunner} has
+   * set them.
+   *
+   * @see TestDataflowRunner#fromOptions(org.apache.beam.sdk.options.PipelineOptions)
+   */
+  private static void runTestPipeline(Pipeline p) {
+    PCollection<Integer> ints = p.apply(Create.of(1, 2, 3));
+
+    final PCollectionView<Integer> singletonSum =
+        ints.apply(Sum.integersGlobally())
+            .apply(View.<Integer>asSingleton());
+
+    PCollection<String> sumPlusValue =
+        ints.apply(ParDo.of(new DoFn<Integer, String>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            Integer sumPlusValue = c.element() + c.sideInput(singletonSum);
+            c.output(sumPlusValue.toString());
+          }
+        }).withSideInputs(singletonSum));
+
+    PAssert.that(sumPlusValue).containsInAnyOrder("7", "8", "9");
+
+    p.run().waitUntilFinish();
+  }
+
+  private static String ensureSlash(String input) {
+    if (!input.endsWith("/")) {
+      return input + '/';
+    }
+    return input;
+  }
+
+  private static String ensureNoSlash(String input) {
+    if (input.endsWith("/")) {
+      return input.substring(0, input.length() - 1);
+    }
+    return input;
+  }
+
+  @Test
+  public void testWithTrailingSlash() throws Exception {
+    TestDataflowPipelineOptions options =
+        TestPipeline.testingPipelineOptions().as(TestDataflowPipelineOptions.class);
+    Pipeline p = Pipeline.create(options);
+
+    options.setTempLocation(ensureSlash(options.getTempLocation()));
+    options.setGcpTempLocation(ensureSlash(options.getGcpTempLocation()));
+    options.setStagingLocation(ensureSlash(options.getStagingLocation()));
+    runTestPipeline(p);
+
+    // Assert that the options were not changed, at least visibly.
+    assertThat(options.getTempLocation(), equalTo(ensureSlash(options.getTempLocation())));
+    assertThat(options.getGcpTempLocation(), equalTo(ensureSlash(options.getGcpTempLocation())));
+    assertThat(options.getStagingLocation(), equalTo(ensureSlash(options.getStagingLocation())));
+  }
+
+  @Test
+  public void testWithNoTrailingSlash() throws Exception {
+    TestDataflowPipelineOptions options =
+        TestPipeline.testingPipelineOptions().as(TestDataflowPipelineOptions.class);
+    Pipeline p = Pipeline.create(options);
+
+    options.setTempLocation(ensureNoSlash(options.getTempLocation()));
+    options.setGcpTempLocation(ensureNoSlash(options.getGcpTempLocation()));
+    options.setStagingLocation(ensureNoSlash(options.getStagingLocation()));
+    runTestPipeline(p);
+
+    // Assert that the options were not changed, at least visibly.
```

Comment:
This is probably too strict a check - they should point at the same location, but need not be equivalent literals.

I've filed https://issues.apache.org/jira/browse/BEAM-1210 - which will cause these assertions to not be particularly meaningful